### ISSUE TITLE
Token Validation API

### ISSUE THREAD

**Issue by schnittstabil** [lcobucci/jwt#72]
In my opinion, the current Validation API can be greatly improved:

``` php
$data = new ValidationData();
…
$token->validate($data) // returns bool
```
1. returning `true` resp. `false` isn't very useful; returning a list of _constraint violations_ (i.e. Errors or similar) would be more helpful, e.g. like [Symfony and Java](https://github.com/symfony/validator) do.
2. token objects should not be responsible for validation; this violates many design principles (e.g.&nbsp;[SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle)): _Changes_ in the validation process –like I suggest under item 1. – causes _changes_ in the `Token` class. IMO, using some `Validator` class would be a much better option.

**Comment by lcobucci** [comment]
@schnittstabil some tradeoffs were made and when I built the `Token#validate()` I just need to know either if the token is valid or not (the violation list has not adding much value).

But you're right in your points and the next major release (v4) is a great opportunity to fix that. Could you send us a PR?

**Comment by schnittstabil** [comment]
Sorry, I have no time for it, but I'll remember this issue when I'm less busy.

**Comment by dannydorfel** [comment]
@lcobucci I have done some work on nr 2. Creating a validator class with a validate(Token $token, ValidationData $data). Will send in the PR this evening after fixing the tests.

I would also like to suggest the validator raises an Exception containing the errors. Since an invalid token could be potentially dangerous for the security of the system, the validate should not return false. The developer should be aware of the invalid token and be forced to act accordingly.

**Comment by lcobucci** [comment]
@dannydorfel it would be awesome to see your PR. 

I left the decision of raising an exception to the user and I think it makes things simpler in some cases however I see the benefits of that. 

Some validation libraries offers different methods to handle that thing (ex: `validate` and `assert`). 

Right now I'm not too sure what would be the best way to go...

**Comment by steverhoades** [comment]
I ran into a similar issue with wanting to validate the id token.  I've added my own token validator chain that seems to be working ok at the moment, still a WIP. 

https://github.com/steverhoades/oauth2-openid-connect-client/tree/master/src/Validator

I think @schnittstabil has a very good point about SRP here.  I'm happy to provide a PR if you see any value in the direction I have gone.

**Comment by lcobucci** [comment]
@steverhoades we decided to go on [this direction](https://github.com/lcobucci/jwt/compare/refactor/validator-extraction).

It's also a WIP but I believe everything will be much easier.

**Comment by steverhoades** [comment]
@lcobucci I definitely like this approach better.  I do have a few comments though.

I am not a huge fan of the Validation\Result return value of the validate function.
(https://github.com/lcobucci/jwt/compare/refactor/validator-extraction#diff-d06653258c3297ab87279df99a57ca7dR34)

This leads to code like this:
```php
$result = $validator->validate(...);
if ($result->hasErrors()) {
    $errors = $result->getErrors();
}
```
Verses this:
```php
if (false === $validator->validate(...)) {
    $errors = $validator->getErrors();
}
```

I'm also not sure how I feel about throwing an exception from each of the constraints and storing the resulting exception instance on the violations array (https://github.com/lcobucci/jwt/compare/refactor/validator-extraction#diff-d06653258c3297ab87279df99a57ca7dR45).  I'll agree that it saves a call to a getError() method on the constraint, but why not then save the $e->getMessage().  

Oh, and thank you for this wonderful library!

**Comment by lcobucci** [comment]
@steverhoades that validation result actually is going to be changed 😄 

The idea is to be something like:

```php
try {
    $validator->validate($token, ...$constraints);
} catch (InvalidTokenException $e) {
    // and to get the violations simply $e->violations().
}
```

**Comment by lcobucci** [comment]
> Oh, and thank you for this wonderful library!

Glad to help! 🍻

**Comment by lcobucci** [comment]
Finally fixed on #129
